# ELE: Elucidating tissue-specific lncRNA essentiality by using deep graph neural networks for heterogenous genomic and epigenomic information aggregation

In this work, we proposed a framework to identify essential lncRNAs by taking advantage of the topological feature of the lncRNA-protein-protein heterogeneous network and genomic and epigenomic features. We used genomic and epigenomic features to annotate nodes at lncRNA-protein-protein interaction network and introduced the HinSAGE algorithm to learn node representation for lncRNAs in the lncRNA-protein-protein heterogeneous network. We named this method as ELE.

---
## Project Structure

```
ELE-main/
├── data/   
│   ├── raw/                    # Data before preprocess
│   ├── benchmark/              # Human and mouse benchmark datasets
│   ├── LPI/                    # Human and mouse LPI networks
│   ├── PPI/                    # Human and mouse PPI networks
│   └── LPPI/                   # Human and mouse LPPI networks
├── reference_lncRNA/ 
│   ├── human/                  # Reference data for human lncRNA genes
│   ├── mouse/                  # Reference data for mouse lncRNA genes
│   ├── reference_genome/       # Reference genome sequence for human and mouse
│   └── preprocess/             # Scripts for process 
├── process/                    # Data preprocessing scripts
│   ├── bencmark/               # Building the benchmark dataset
│   └── construct_lppi/         # Constructing the LPPI network
├── features/                   # Data for node annotation
│   ├── conservation/           # Evolutionary conservation data
│   ├── epigenomic/             # Epigenomic feature data from ENCODE
│   └── protein/                # Protein-coding gene related data
├── annotate/                   # Scripts for annotating node features
│   ├── human/                  # Annotate node for human LPPI network
│   └── mouse/                  # Annotate node for mouse LPPI network
├── HinSAGE/                    # Training HinSAGE model
├── classifier/                 # Classifier training scripts
│   ├── SVM/                    # SVM-based cross-validation and prediction
│   └── MLP/                    # MLP-based cross-validation and prediction
├── results/                    # Prediction results generated by ELE
└── experiment/                 # Scripts for experimental procedures
```

---
## Installation 
```
# Clone the repository
git clone https://github.com/swt1024/ELE.git
cd ELE

# Create and activate conda environment (recommended)
conda create -n ELE python==3.8
conda activate ELE

# Install dependencies with CUDA 11.2 support
conda install -c stellargraph stellargraph
conda install -c bioconda bedtools
conda install -c conda-forge pybigwig
pip install -r requirements.txt
```
---
## Workflow of ELE

### 1. Data collection and preprocessing
#### Step1: Download and preprocess the reference lncRNA gene-related data.

1. Download the following files into the corresponding folders:
	`reference_lncRNA/human/gtf`:
	- `NONCODEv5_human_hg38_lncRNA.gtf` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_human_hg38_lncRNA.gtf.gz)  
	- `NONCODEv6_human_hg38_lncRNA.gtf` – [Download](http://www.noncode.org/datadownload/NONCODEv6_human_hg38_lncRNA.gtf.gz)  
	- `ensembl/`: Run `./download.sh` to download multiple GTF files from Ensembl, including: `v76`, `v78`, `v80`, `v84`, `v87`, `v93`, `v97`, `v104`, `v106`, `v107`, `v108`, `v109`, `v110`, `v111`, `v112`, `v113`. Example (Release 113): [Homo_sapiens.GRCh38.113.gtf.gz](https://ftp.ensembl.org/pub/release-113/gtf/homo_sapiens/Homo_sapiens.GRCh38.113.gtf.gz) More versions available at: [Ensembl FTP Repository](https://ftp.ensembl.org/pub/) 
	
	`reference_lncRNA/human/bed`:
	- `NONCODEv6_hg38.lncAndGene.bed` – [Download](http://www.noncode.org/datadownload/NONCODEv6_hg38.lncAndGene.bed.gz)  
	- `NONCODEv5_hg38.lncAndGene.bed` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_hg38.lncAndGene.bed.gz)

	`reference_lncRNA/human/fasta`
	- `NONCODEv6_human.fa` – [Download](http://www.noncode.org/datadownload/NONCODEv6_human.fa.gz)
	- `NONCODEv5_human.fa` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_human.fa.gz)
	- `ensembl/`: Run `./download.sh` to download multiple FASTA files from Ensembl, including: `v76`, `v78`, `v80`, `v84`, `v87`, `v93`, `v97`, `v104`, `v106`, `v107`, `v108`, `v109`, `v110`, `v111`, `v112`, `v113`. Example (Release 113): [Homo_sapiens.GRCh38.ncrna.fa.gz](https://ftp.ensembl.org/pub/release-113/fasta/homo_sapiens/ncrna/Homo_sapiens.GRCh38.ncrna.fa.gz) More versions available at: [Ensembl FTP Repository](https://ftp.ensembl.org/pub/)

	`reference_lncRNA/mouse/gtf`:
	- `NONCODEv5_mouse_mm10_lncRNA.gtf` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_mouse_mm10_lncRNA.gtf.gz)  
	- `ensembl/`: Run `./download.sh` to download multiple GTF files from Ensembl, including v84–v100. Example (Release 100): [Mus_musculus.GRCm38.100.gtf.gz](https://ftp.ensembl.org/pub/release-100/gtf/mus_musculus/Mus_musculus.GRCm38.100.gtf.gz) More versions available at: [Ensembl FTP Repository](https://ftp.ensembl.org/pub/)
	
	`reference_lncRNA/mouse/bed`:
	- `NONCODEv6_mm10.lncAndGene.bed` – [Download](http://www.noncode.org/datadownload/NONCODEv6_mm10.lncAndGene.bed.gz)  
	- `NONCODEv5_mm10.lncAndGene.bed` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_mm10.lncAndGene.bed.gz)
	
	`reference_lncRNA/mouse/fasta`
	- `NONCODEv5_mouse.fa` – [Download](http://v5.noncode.org/datadownload/NONCODEv5_mouse.fa.gz)
	- `NONCODEv6_mouse.fa` – [Download](http://www.noncode.org/datadownload/NONCODEv6_mouse.fa.gz)
	- `ensembl/`: Run `./download.sh` to download multiple FASTA files from Ensembl, including v84–v100. Example (Release 100): [Mus_musculus.GRCm38..ncrna.fa.gz](https://ftp.ensembl.org/pub/release-100/fasta/mus_musculus/ncrna/Mus_musculus.GRCm38.ncrna.fa.gz) More versions available at: [Ensembl FTP Repository](https://ftp.ensembl.org/pub/)
	
	`reference_lncRNA/reference_genome`
	- `GRCh38.p14.genome.fa`  – [Download](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_47/GRCh38.p14.genome.fa.gz)
	- `GRCm38.p6.genome.fa`   – [Download](https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_mouse/release_M25/GRCm38.p6.genome.fa.gz)

2. Run the following command line instructions for preprocessing in the `reference_lncRNA/preprocess/` directory:
	```
	# Preprocess NONCODE BED files
	./pro_bed.sh
	
	# Convert Ensembl GTF files into BED files
	python gtf2bed.py ../human/gtf/ensembl ../human/bed/ensembl
	python gtf2bed.py ../mouse/gtf/ensembl ../mouse/bed/ensembl
	
	# Generate mappings between lncRNA genes and transcripts from GTF files
	python get_trans.py ../human/gtf ../human/transcript
	python get_trans_ensembl.py ../human/gtf/ensembl ../human/transcript/ensmebl
	python get_trans.py ../mouse/gtf ../mouse/transcript
	python get_trans_ensembl.py ../mouse/gtf/ensembl ../mouse/transcript/ensmebl
	
	# Preprocess FASTA files
	python process_fasta.py ../human/fasta ../human/fasta
	python process_fasta.py ../human/fasta/ensembl ../human/fasta/processed_ensembl
	python process_fasta.py ../mouse/fasta ../mouse/fasta
	python process_fasta.py ../mouse/fasta/ensembl ../mouse/fasta/processed_ensembl
	```
#### Step2: Download LPI, PPI, and essential lncRNA data.
1. Download the following data to `data/raw`:
	- `esslnc.csv`: Essential lncRNA gene data from the dbEssLnc2.0 database. -[Download](https://esslnc.pufengdu.org/v2/data/esslnc.xlsx)
	- `lncRNA_interaction.txt`: Raw LPI data from the NPInter v5.0 database.  -[Download](http://bigdata.ibp.ac.cn/npinter5/download/file/lncRNA_interaction.txt.gz)
	  [Download](http://bigdata.ibp.ac.cn/npinter5/download/file/lncRNA_interaction.txt.gz)
	- `BIOGRID-Homo_sapiens-4.4.248.tab3.txt`: PPI data from the BIOGRID database. 
	  [Download shared ZIP](https://downloads.thebiogrid.org/File/BioGRID/Release-Archive/BIOGRID-4.4.248/BIOGRID-ORGANISM-4.4.248.tab3.zip)
	- `BIOGRID-Mus_musculus-4.4.248.tab3.txt`: PPI data from the BIOGRID database.  
	  [Download shared ZIP](https://downloads.thebiogrid.org/File/BioGRID/Release-Archive/BIOGRID-4.4.248/BIOGRID-ORGANISM-4.4.248.tab3.zip)

#### Step3: Download and preprocess the data used for annotating nodes.
`features/conservation`:
Run the `download.sh` script to download the phylop and phastCons scores for both human and mouse genomes in BigBed format from UCSC.
`features/epigenomic`:
Run `python download.py human` and `python download.py mouse` to download epigenomic peak data in BigBed format from ENCODE for various tissues. Specific download links can be found in the respective species folder and tissue folder under the `download.sh` file.
`features/protein/mouse`:
1. Run the `download.py` script to download mouse gene expression data for multiple tissues from MGI. The script uses the following file:
	- `MGI_filter_conditions.csv`: Contains file names and the corresponding filter conditions for each dataset.
2. Download the `mgi.gaf` file, which contains the Gene Ontology (GO) term annotations for mouse genes. - [Download](https://current.geneontology.org/annotations/mgi.gaf.gz)
3. Run the `filter.py` script to filter the expression data files based on the conditions specified in the `MGI_filter_conditions.csv` file.
`features/protein/human`:
1. Download `GTEx_Analysis_2022-06-06_v10_RNASeQCv2.4.2_gene_median_tpm.gct`, which contains data of the expression levels of human protein-coding genes from the GTEx database. - [Download](https://storage.googleapis.com/adult-gtex/bulk-gex/v10/rna-seq/GTEx_Analysis_v10_RNASeQCv2.4.2_gene_median_tpm.gct.gz)
2. Run the `filter.py` script to filter the expression data files, obtain `filtered_exp.csv`.
3. Download GO term data for human protein-coding genes `go.csv` from Ensembl BioMart.

### 2. Construct lncRNA-protein-protein heterogeneous network

1. Run `process/construct_lppi/get_lpi.py` to preprocess the NPInter LPI data and construct a weighted LPI network.
2. Run `process/construct_lppi/get_lppi.py` to filter the BioGRID PPI data and generate the LPPI network.

### 3. Node feature annotation for the LPPI network

**human**
1. Run the `annotate/human/lncRNA_feature_annotate.ipynb` notebook to annotate features for all lncRNA nodes in the human LPPI network.
2. Run the `annotate/human/protein_feature_annotate.ipynb` notebook to annotate features for all protein nodes in the human LPPI network.

**mouse**
1. Run the `annotate/mouse/lncRNA_feature_annotate.ipynb` notebook to annotate features for all lncRNA nodes in the mouse LPPI network.
2. Run the `annotate/mouse/protein_feature_annotate.ipynb` notebook to annotate features for all protein nodes in the mouse LPPI network.

### 4. Heterogeneous representation learning

You can train the model using the provided shell script:
```
python train.py --layer_sizes 64 256 --samples_num 20 25 --lncRNA_nodes_file "../annotate/human/valid_heart_annotation.csv" --protein_nodes_file "../annotate/human/transformed_protein_annotation.csv" --lppi_file "../annotate/human/valid_inter.csv" --embedding_save_path "./human/lncRNA_embeddings_heart.csv"
```
|                        |                                |
|------------------------|--------------------------------|
|     **Parameter**      |         **Description**        |
|    `--layer_sizes`     |   Layer sizes for the model.   |
|    `--samples_num`     |  Number of samples per layer.  |
|  `--lncRNA_nodes_file` | Path to the lncRNA nodes file. |
| `--protein_nodes_file` |Path to the protein nodes file. |
|    `--lppi_file`       |    Path to the LPPI file.      |
|`--embedding_save_path` |Path to save the embedding file.|

The following are the optimal parameters used on the datasets in this study:

| dataset | The number of sampled neighboring nodes | The dimensions of the hidden layers |
| ------- | :-------------------------------------: | :---------------------------------: |
| Human   |                 (20,25)                 |              (64,256)               |
| Mouse   |                 (10,15)                 |              (32, 64)               |

If you want to train the model on your own dataset, you need to provide the following files:

1. Nodes File (--lncRNA_nodes_file and --protein_nodes_file):
- CSV files containing features for lncRNAs and proteins.
- Must include a lncRNA_id column (for lncRNAs) and a protein_id column (for proteins).
2. LPPI File (--lppi_file):
- A CSV file that defines interactions between lncRNAs and proteins.
- Must include lncRNA_id and protein_id columns.

**Note:**  
You can run `./HinSAGE/tune.py` to generate node embeddings for tuning model parameters.  
You can then use the first two code cells in `classifier/SVM/svm.ipynb` to evaluate the model's performance under different parameter settings.

### 5. Construct benchmark dataset
1. Run `process/benchmark/get_trans.ipynb` to obtain transcript information for each lncRNA gene and filter out transcripts longer than 20,000 nt.
2. In the `process/benchmark` folder, run `./cal_MFE.sh ./human/transcript_sequences.fasta ./human/trans_MFE.csv` and `./cal_MFE.sh ./mouse/transcript_sequences.fasta ./mouse/trans_MFE.csv` to compute the MFE for each transcript.
3. Run `construct_benchmark.ipynb` to calculate the GIC score and obtain the positive and negative samples.

### 6. Supervised machine learning

#### SVM model

| dataset |   C   | kernel function |
| :-----: |  :-:  | :-------------: |
|  Human  |  100  |     linear      |
|  Mouse  |  10   |     linear      |

The table above shows the parameters used by the SVM model under different datasets.
For example, users can run the forth code cell in `svm.ipynb`, input `data/benchmark/human/ess_lpi.csv`, `data/benchmark/human/noness_lpi.csv` and `HinASGE/human/lncRNA_embeddings_heart`, and get accuracy, precision and other performance indicators.
You can run the fifth code cell in `svm.ipynb`, input `data/benchmark/human/ess_lpi.csv`, `data/benchmark/human/noness_lpi.csv` and `HinASGE/human/lncRNA_embeddings_heart.csv`, and get predicted results.

#### MLP model

| dataset | The number of neurons in the hidden layer |
| :-----: | :---------------------------------------: |
|  Human  |                 (128,64)                  |
|  Mouse  |                 (32, 16)                  |

The table above shows the parameters used by the MLP model under different datasets.

The prediction results of SVM and MLP will be stored in the `results/human` and `results/mouse` directories. The `results` folder contains the prediction outcomes of SVM and MLP across different tissues. It also includes the intersection of the essential lncRNA genes predicted by the two classifiers for each tissue, as well as the union of essential lncRNA genes predicted across multiple tissues for each species. These results can be used for subsequent experimental analyses, including but not limited to the experiments in `experiment`.