{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "334f5c3c",
   "metadata": {},
   "source": [
    "### Calculate GIC score"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6b53fce0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Transcripts: 100%|██████████| 163300/163300 [02:02<00:00, 1330.57it/s]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GIC scores have been calculated and saved successfully.\n",
      "Output file is located at: ./human/sorted_GIC_score.csv\n"
     ]
    }
   ],
   "source": [
    "# Human\n",
    "\n",
    "import math\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "# record all triplet combinations\n",
    "BASES = ['a', 't', 'c', 'g']\n",
    "TRIPLETS = []\n",
    "for B1 in BASES:\n",
    "    for B2 in BASES:\n",
    "        for B3 in BASES:\n",
    "            TRIPLETS.append(B1+B2+B3)\n",
    "\n",
    "mers = {} #construct a list of triplets that are initialized to 0\n",
    "\n",
    "# obtain the corresponding relationship between lncRNA and transcript\n",
    "def readlnc_transID(filename):\n",
    "    lnc_trans = pd.read_csv(filename, header=None, names=['lncRNA_id', 'transcript_id'])\n",
    "    #Create the mapping relationship between genes and transcripts\n",
    "    lnc_trans_mapping = lnc_trans.groupby('lncRNA_id')['transcript_id'].aggregate(list).to_dict()\n",
    "\n",
    "    return lnc_trans_mapping\n",
    "\n",
    "# Get the secondary structure minimum free energy of each transcript of lncRNA\n",
    "def readlnc_mfe(filename):\n",
    "    trans_mfe = pd.read_csv(filename)\n",
    "    #Create the mapping relationship between transcripts and MFE\n",
    "    trans_mfe_mapping = pd.Series(trans_mfe['MFE'].values, index=trans_mfe['transcript_id']).to_dict()\n",
    "\n",
    "    return trans_mfe_mapping\n",
    "\n",
    "# Get transcript sequence\n",
    "def readfasta(filename):\n",
    "    f = open(filename,'r')\n",
    "    res = {} #Preserve transcripts and their sequence\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            line = line.strip()\n",
    "            ID = line.split('>',2)[1]\n",
    "            res[ID] = ''\n",
    "        else:\n",
    "            res[ID] += line\n",
    "    return res\n",
    "\n",
    "#sliding window\n",
    "def slidingWindow(seq, l, win, step=1): \n",
    "    length = l\n",
    "    mod = divmod((length-win), step)[1]\n",
    "    if (win >= length):\n",
    "        return seq\n",
    "    else:\n",
    "        start = 0\n",
    "        end = win\n",
    "        fragments = []\n",
    "        while (len(seq[start:end]) == win):\n",
    "            fragments.append(seq[start:end])\n",
    "            start += step\n",
    "            end += step\n",
    "        if (mod > 0):\n",
    "            fragments.append(seq[(length-win):])\n",
    "        return fragments\n",
    " \n",
    "#calculate the frequency of the triplet\n",
    "def stat3mer(seq, l):   \n",
    "    freq = {}\n",
    "    for item in TRIPLETS:\n",
    "        mers[item] = 0\n",
    "    num3mer = float(l-2)\n",
    "    all3mer = slidingWindow(seq, l, win=3, step=1)\n",
    "    for i in set(TRIPLETS):\n",
    "        mers[i] = all3mer.count(i)\n",
    "    for triplet in TRIPLETS:\n",
    "        freq[triplet] = mers[triplet]/num3mer\n",
    "    return freq\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    lnc_trans_path = './human/filtered_lnc_trans.csv'  \n",
    "    MFE_path = './human/trans_MFE.csv'\n",
    "    seq_path = './human/transcript_sequences.fasta'  \n",
    "\n",
    "    lnc_transID = readlnc_transID(lnc_trans_path)\n",
    "    trans_mfe = readlnc_mfe(MFE_path)\n",
    "    trans_seq = readfasta(seq_path)\n",
    "\n",
    "    # parameter Settings of GIC\n",
    "    LRMODEL_FEATURES_7 = ['intercept', 'length', 'mfe/L', 'cga', 'gcg', 'tcg', 'acg', 'tca']\n",
    "    LRMODEL_COEFS_7 = [0.7417, 2.612e-04, 4.295, 48.66, 15.64, 76.23, -1.113, -60.29]\n",
    "    LRMODEL_7 = dict(zip(LRMODEL_FEATURES_7, LRMODEL_COEFS_7))\n",
    "\n",
    "    #calculated the eigenvalues of each transcript\n",
    "    features_trans = {}  #save the features of each transcript\n",
    "    for id, seq in tqdm(trans_seq.items(), desc=\"Processing Transcripts\"):\n",
    "        if id not in trans_mfe:\n",
    "            continue\n",
    "        feature = {}\n",
    "        seq = seq.replace('\\n','')\n",
    "        seq = seq.lower()\n",
    "        length = len(seq)\n",
    "        feature['trans'] = id\n",
    "        feature['length'] = length\n",
    "        feature['mfe/L'] = trans_mfe[id]/length\n",
    "        freq = stat3mer(seq, length)\n",
    "        for item in TRIPLETS:\n",
    "            feature[item] = freq[item]\n",
    "        # tmp = LRMODEL_7['intercept']+sum([feature[_]*LRMODEL_7[_] for _ in LRMODEL_FEATURES_7[1:]])\n",
    "        # gic = math.exp(tmp)/(math.exp(tmp)+1)\n",
    "        # feature['GIC_7'] = gic\n",
    "        features_trans[id] = feature\n",
    "    \n",
    "    #calculated the eigenvalues of each lncRNA\n",
    "    lncRNA_GIC_score = {}  ##save the GIC score of each lncRNA\n",
    "    for lnc,transcripts in lnc_transID.items():\n",
    "        feature = {} #save the features of lncRNA gene\n",
    "        feature['length'] = 0 # initialize the value of each feature to 0\n",
    "        feature['mfe/L'] = 0\n",
    "        for item in TRIPLETS:\n",
    "            feature[item] = 0\n",
    "    \n",
    "        for trans in transcripts:\n",
    "            if trans not in features_trans:\n",
    "                feature = {}\n",
    "                break\n",
    "            else:\n",
    "                tranfeature = features_trans[trans]\n",
    "                feature['length'] += tranfeature['length']\n",
    "                feature['mfe/L']  += tranfeature['mfe/L']\n",
    "                for item in TRIPLETS:\n",
    "                    feature[item] += tranfeature[item]\n",
    "        if len(feature)>0:\n",
    "            feature['lncRNA_id'] = lnc\n",
    "            feature['length'] = feature['length']/len(transcripts)\n",
    "            feature['mfe/L'] = feature['mfe/L']/len(transcripts)\n",
    "            for item in TRIPLETS:\n",
    "                feature[item] =  feature[item]/len(transcripts)\n",
    "            tmp = LRMODEL_7['intercept']+sum([feature[_]*LRMODEL_7[_] for _ in LRMODEL_FEATURES_7[1:]])\n",
    "            gic = math.exp(tmp)/(math.exp(tmp)+1)\n",
    "            feature['GIC_score'] = gic\n",
    "            lncRNA_GIC_score[lnc] = feature['GIC_score']\n",
    "\n",
    "    # sort all LncRNAs according to GIC scores\n",
    "    lncRNA_GIC_score = sorted(lncRNA_GIC_score.items(),key=lambda d:d[1],reverse=True)\n",
    "    \n",
    "    # Convert the lncRNA GIC scores to a DataFrame\n",
    "    lncRNA_scores_df_sorted = pd.DataFrame(lncRNA_GIC_score, columns=['lncRNA_id', 'GIC_score'])\n",
    "\n",
    "    # Write the sorted results to a CSV file\n",
    "    sorted_GIC_path = './human/sorted_GIC_score.csv'\n",
    "    lncRNA_scores_df_sorted.to_csv(sorted_GIC_path, index=False)\n",
    "\n",
    "    print(\"GIC scores have been calculated and saved successfully.\")\n",
    "    print(f\"Output file is located at: {sorted_GIC_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ab19220c",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Processing Transcripts: 100%|██████████| 56374/56374 [01:23<00:00, 675.32it/s] \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GIC scores have been calculated and saved successfully.\n",
      "Output file is located at: ./mouse/sorted_GIC_score.csv\n"
     ]
    }
   ],
   "source": [
    "# Mouse\n",
    "\n",
    "import math\n",
    "import pandas as pd\n",
    "from tqdm import tqdm\n",
    "\n",
    "\n",
    "# record all triplet combinations\n",
    "BASES = ['a', 't', 'c', 'g']\n",
    "TRIPLETS = []\n",
    "for B1 in BASES:\n",
    "    for B2 in BASES:\n",
    "        for B3 in BASES:\n",
    "            TRIPLETS.append(B1+B2+B3)\n",
    "\n",
    "mers = {} #construct a list of triplets that are initialized to 0\n",
    "\n",
    "# obtain the corresponding relationship between lncRNA and transcript\n",
    "def readlnc_transID(filename):\n",
    "    lnc_trans = pd.read_csv(filename, header=None, names=['lncRNA_id', 'transcript_id'])\n",
    "    #Create the mapping relationship between genes and transcripts\n",
    "    lnc_trans_mapping = lnc_trans.groupby('lncRNA_id')['transcript_id'].aggregate(list).to_dict()\n",
    "\n",
    "    return lnc_trans_mapping\n",
    "\n",
    "# Get the secondary structure minimum free energy of each transcript of lncRNA\n",
    "def readlnc_mfe(filename):\n",
    "    trans_mfe = pd.read_csv(filename)\n",
    "    #Create the mapping relationship between transcripts and MFE\n",
    "    trans_mfe_mapping = pd.Series(trans_mfe['MFE'].values, index=trans_mfe['transcript_id']).to_dict()\n",
    "\n",
    "    return trans_mfe_mapping\n",
    "\n",
    "# Get transcript sequence\n",
    "def readfasta(filename):\n",
    "    f = open(filename,'r')\n",
    "    res = {} #Preserve transcripts and their sequence\n",
    "    for line in f:\n",
    "        if line.startswith('>'):\n",
    "            line = line.strip()\n",
    "            ID = line.split('>',2)[1]\n",
    "            res[ID] = ''\n",
    "        else:\n",
    "            res[ID] += line\n",
    "    return res\n",
    "\n",
    "#sliding window\n",
    "def slidingWindow(seq, l, win, step=1): \n",
    "    length = l\n",
    "    mod = divmod((length-win), step)[1]\n",
    "    if (win >= length):\n",
    "        return seq\n",
    "    else:\n",
    "        start = 0\n",
    "        end = win\n",
    "        fragments = []\n",
    "        while (len(seq[start:end]) == win):\n",
    "            fragments.append(seq[start:end])\n",
    "            start += step\n",
    "            end += step\n",
    "        if (mod > 0):\n",
    "            fragments.append(seq[(length-win):])\n",
    "        return fragments\n",
    " \n",
    "#calculate the frequency of the triplet\n",
    "def stat3mer(seq, l):   \n",
    "    freq = {}\n",
    "    for item in TRIPLETS:\n",
    "        mers[item] = 0\n",
    "    num3mer = float(l-2)\n",
    "    all3mer = slidingWindow(seq, l, win=3, step=1)\n",
    "    for i in set(TRIPLETS):\n",
    "        mers[i] = all3mer.count(i)\n",
    "    for triplet in TRIPLETS:\n",
    "        freq[triplet] = mers[triplet]/num3mer\n",
    "    return freq\n",
    "\n",
    "\n",
    "if __name__ == '__main__':\n",
    "\n",
    "    lnc_trans_path = './mouse/filtered_lnc_trans.csv'  \n",
    "    MFE_path = './mouse/trans_MFE.csv'\n",
    "    seq_path = './mouse/transcript_sequences.fasta'  \n",
    "\n",
    "    lnc_transID = readlnc_transID(lnc_trans_path)\n",
    "    trans_mfe = readlnc_mfe(MFE_path)\n",
    "    trans_seq = readfasta(seq_path)\n",
    "\n",
    "    # parameter Settings of GIC\n",
    "    LRMODEL_FEATURES_7 = ['intercept', 'length', 'mfe/L', 'cga', 'gcg', 'tcg', 'acg', 'tca']\n",
    "    LRMODEL_COEFS_7 = [0.1625, 2.638e-04, 2.194, 19.88, 37.59, 50.37, 35.44, -64.66]\n",
    "    LRMODEL_7 = dict(zip(LRMODEL_FEATURES_7, LRMODEL_COEFS_7))\n",
    "\n",
    "    #calculated the eigenvalues of each transcript\n",
    "    features_trans = {}  #save the features of each transcript\n",
    "    for id, seq in tqdm(trans_seq.items(), desc=\"Processing Transcripts\"):\n",
    "        if id not in trans_mfe:\n",
    "            continue\n",
    "        feature = {}\n",
    "        seq = seq.replace('\\n','')\n",
    "        seq = seq.lower()\n",
    "        length = len(seq)\n",
    "        feature['trans'] = id\n",
    "        feature['length'] = length\n",
    "        feature['mfe/L'] = trans_mfe[id]/length\n",
    "        freq = stat3mer(seq, length)\n",
    "        for item in TRIPLETS:\n",
    "            feature[item] = freq[item]\n",
    "        # tmp = LRMODEL_7['intercept']+sum([feature[_]*LRMODEL_7[_] for _ in LRMODEL_FEATURES_7[1:]])\n",
    "        # gic = math.exp(tmp)/(math.exp(tmp)+1)\n",
    "        # feature['GIC_7'] = gic\n",
    "        features_trans[id] = feature\n",
    "    \n",
    "    #calculated the eigenvalues of each lncRNA\n",
    "    lncRNA_GIC_score = {}  ##save the GIC score of each lncRNA\n",
    "    for lnc,transcripts in lnc_transID.items():\n",
    "        feature = {} #save the features of lncRNA gene\n",
    "        feature['length'] = 0 # initialize the value of each feature to 0\n",
    "        feature['mfe/L'] = 0\n",
    "        for item in TRIPLETS:\n",
    "            feature[item] = 0\n",
    "    \n",
    "        for trans in transcripts:\n",
    "            if trans not in features_trans:\n",
    "                feature = {}\n",
    "                break\n",
    "            else:\n",
    "                tranfeature = features_trans[trans]\n",
    "                feature['length'] += tranfeature['length']\n",
    "                feature['mfe/L']  += tranfeature['mfe/L']\n",
    "                for item in TRIPLETS:\n",
    "                    feature[item] += tranfeature[item]\n",
    "        if len(feature)>0:\n",
    "            feature['lncRNA_id'] = lnc\n",
    "            feature['length'] = feature['length']/len(transcripts)\n",
    "            feature['mfe/L'] = feature['mfe/L']/len(transcripts)\n",
    "            for item in TRIPLETS:\n",
    "                feature[item] =  feature[item]/len(transcripts)\n",
    "            tmp = LRMODEL_7['intercept']+sum([feature[_]*LRMODEL_7[_] for _ in LRMODEL_FEATURES_7[1:]])\n",
    "            gic = math.exp(tmp)/(math.exp(tmp)+1)\n",
    "            feature['GIC_score'] = gic\n",
    "            lncRNA_GIC_score[lnc] = feature['GIC_score']\n",
    "\n",
    "    # sort all LncRNAs according to GIC scores\n",
    "    lncRNA_GIC_score = sorted(lncRNA_GIC_score.items(),key=lambda d:d[1],reverse=True)\n",
    "    \n",
    "    # Convert the lncRNA GIC scores to a DataFrame\n",
    "    lncRNA_scores_df_sorted = pd.DataFrame(lncRNA_GIC_score, columns=['lncRNA_id', 'GIC_score'])\n",
    "\n",
    "    # Write the sorted results to a CSV file\n",
    "    sorted_GIC_path = './mouse/sorted_GIC_score.csv'\n",
    "    lncRNA_scores_df_sorted.to_csv(sorted_GIC_path, index=False)\n",
    "\n",
    "    print(\"GIC scores have been calculated and saved successfully.\")\n",
    "    print(f\"Output file is located at: {sorted_GIC_path}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cdca003",
   "metadata": {},
   "source": [
    "### Get essential lncRNA samples and non-essential lncRNA samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "e39b00b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "esslnc = pd.read_csv(\"../../data/raw/esslnc.csv\")\n",
    "\n",
    "esslnc = esslnc[[\"Noncode_id\",\"target\",\"gene_name\",\"ensembl_id\",\"Organism\",\"cancer_related\",\"disease_related\",\"vivo\",\"vitro\"]]\n",
    "\n",
    "human_esslnc = esslnc[esslnc['Organism']=='Human']\n",
    "mouse_esslnc = esslnc[esslnc['Organism']=='Mouse']\n",
    "\n",
    "human_esslnc.to_csv('../../data/benchmark/human/human_esslnc.csv',index=False)\n",
    "mouse_esslnc.to_csv('../../data/benchmark/mouse/mouse_esslnc.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5138a5fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Human\n",
    "import pandas as pd\n",
    "\n",
    "# Read the lncRNA data\n",
    "ess_lnc = pd.read_csv(\"../../data/benchmark/human/human_esslnc.csv\")\n",
    "lncRNA = pd.read_csv(\"../../data/LPI/human/lncRNA.csv\")\n",
    "annotation = pd.read_csv(\"../../annotate/human/valid_heart_annotation.csv\")\n",
    "\n",
    "annotated_lncRNA = lncRNA[lncRNA['lncRNA_id'].isin(annotation['lncRNA_id'])].copy()\n",
    "\n",
    "ess_lnc['Noncode_id'] = ess_lnc['Noncode_id'].str.split('.').str[0]\n",
    "def is_essential(id, name):\n",
    "    if id != '-' and any(ess_lnc[col].isin([id]).any() for col in ['Noncode_id', 'ensembl_id']):\n",
    "        return 1\n",
    "    if name != '-' and any(ess_lnc[col].isin([name]).any() for col in ['gene_name', 'target']):\n",
    "        return 1\n",
    "    return 0\n",
    "\n",
    "annotated_lncRNA.loc[:, 'essential'] = annotated_lncRNA.apply(lambda row: is_essential(row['gene_id'], row['gene_name']), axis=1)\n",
    "ess_lnc = annotated_lncRNA[annotated_lncRNA['essential'] == 1]\n",
    "ess_lnc = ess_lnc[['lncRNA_id']]\n",
    "\n",
    "ess_lnc.to_csv('../../data/benchmark/human/ess_lnc.csv', index=False)\n",
    "\n",
    "lnc_GIC = pd.read_csv(\"./human/sorted_GIC_score.csv\")\n",
    "\n",
    "unlabel_lnc = lnc_GIC[~lnc_GIC['lncRNA_id'].isin(ess_lnc['lncRNA_id'])]\n",
    "unlabel_lnc = unlabel_lnc[unlabel_lnc['lncRNA_id'].isin(annotated_lncRNA['lncRNA_id'])]\n",
    "\n",
    "ess_counts = ess_lnc.shape[0]\n",
    "\n",
    "noness_lnc = unlabel_lnc.tail(ess_counts)[['lncRNA_id']].reset_index(drop=True)\n",
    "\n",
    "noness_lnc.to_csv(\"../../data/benchmark/human/noness_lnc.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f37be26",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Mouse\n",
    "import pandas as pd\n",
    "\n",
    "# Read the lncRNA data\n",
    "ess_lnc = pd.read_csv(\"../../data/benchmark/mouse/mouse_esslnc.csv\")\n",
    "lncRNA = pd.read_csv(\"../../data/LPI/mouse/lncRNA.csv\")\n",
    "annotation = pd.read_csv(\"../../annotate/mouse/valid_heart_annotation.csv\")\n",
    "\n",
    "annotated_lncRNA = lncRNA[lncRNA['lncRNA_id'].isin(annotation['lncRNA_id'])].copy()\n",
    "\n",
    "ess_lnc['Noncode_id'] = ess_lnc['Noncode_id'].str.split('.').str[0]\n",
    "def is_essential(id, name):\n",
    "    if id != '-' and any(ess_lnc[col].isin([id]).any() for col in ['Noncode_id', 'ensembl_id']):\n",
    "        return 1\n",
    "    if name != '-' and any(ess_lnc[col].isin([name]).any() for col in ['gene_name', 'target']):\n",
    "        return 1\n",
    "    return 0\n",
    "\n",
    "annotated_lncRNA.loc[:, 'essential'] = annotated_lncRNA.apply(lambda row: is_essential(row['gene_id'], row['gene_name']), axis=1)\n",
    "\n",
    "ess_lnc = annotated_lncRNA[annotated_lncRNA['essential'] == 1]\n",
    "ess_lnc = ess_lnc[['lncRNA_id']]\n",
    "\n",
    "ess_lnc.to_csv('../../data/benchmark/mouse/ess_lnc.csv', index=False)\n",
    "\n",
    "lnc_GIC = pd.read_csv(\"./mouse/sorted_GIC_score.csv\")\n",
    "\n",
    "unlabel_lnc = lnc_GIC[~lnc_GIC['lncRNA_id'].isin(ess_lnc['lncRNA_id'])]\n",
    "unlabel_lnc = unlabel_lnc[unlabel_lnc['lncRNA_id'].isin(annotated_lncRNA['lncRNA_id'])]\n",
    "\n",
    "ess_counts = ess_lnc.shape[0]\n",
    "\n",
    "noness_lnc = unlabel_lnc.tail(ess_counts)[['lncRNA_id']].reset_index(drop=True)\n",
    "\n",
    "noness_lnc.to_csv(\"../../data/benchmark/mouse/noness_lnc.csv\", index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "esslnc",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
